content/blog/2022/2022-08-03-feature-flags-in-nodejs-react-client-with-launchdarkly.md

```diff
@@ -0,0 +1,305 @@
+
+## Create a Simple React App
+
+Let’s start by creating a simple React UI application. We will build this using one of the most popular npm libraries, [create-react-app](https://create-react-app.dev/). It is one of the simplest methods to start a new React project, making it a great option for both serious, large-scale apps and your own, individual projects.
+
+We will use the tool npx which gives us the ability to use the `create-react-app` package without having to first install it on our computer, which is very convenient. It also ensures that we are always using the latest version of `create-react-app`.
+
+```bash
+npx create-react-app nodejs-react-feature-flag-launchdarkly
+```
+
+This will create a folder with the name “nodejs-react-feature-flag-launchdarkly” and automatically install all the required packages inside that folder. The folder tree would look something like below:
+
+```bash
+nodejs-react-feature-flag-launchdarkly
+├── README.md
+├── node_modules
+├── package.json
+├── .gitignore
+├── public
+└── src
+```
+
+Now we can simply run our app and see a dev server starting up and opening a UI with the React logo:
+
+```bash
+npm start
+```
+
+The development server will start up at http://localhost:3000 and we can see the home page.
+
+{{% image alt="React UI" src="images/posts/nodejs-frontend-launchdarkly/React_Logo.png" %}}
+
+Next we will add the [launchdarkly-js-client-sdk](https://github.com/launchdarkly/js-client-sdk) package to use client-side methods to fetch the feature flag variations:
+
+```bash
+npm install launchdarkly-js-client-sdk
+```
+
+
+
+## String-based Feature Flag to Hide/Display Content on the Fly
+
+In our first use-case, let’s try to simply define a string feature flag in LaunchDarkly UI. We will simply list some users in the React UI. If any of the users matches the user name defined in the feature flag variation, then we will encrypt that user name in the UI. It would be pretty simple but enough to understand how we can enable/disable a feature in web UI.
+
+The feature flag defined in LaunchDarkly UI would look something like this:
+
+{{% image alt="String Feature Flag LaunchDarkly" src="images/posts/nodejs-frontend-launchdarkly/Hide_User.png" %}}
+
+By default, flags are only available to server-side SDKs. When we create a flag, we can choose to expose the flag to SDKs which use client-side IDs, SDKs which use mobile keys, or both. If we are using a client-side or mobile SDK, we must expose our feature flags in order for the client-side or mobile SDKs to evaluate them. As per the [documentation](https://docs.launchdarkly.com/home/getting-started/feature-flags#making-flags-available-to-client-side-and-mobile-sdks), in case of React UI or any Javascript UI, we need to enable *“SDKs using Client-side ID”* option. So in this case, we will fetch the *Client-Side ID* from LaunchDarkly Projects page and define it as part of the code. We should also make sure that the targeting option is enabled for the flag.
```

Comment:
This paragraph would be great in a separate section titled "Differences Between Client-side and Server-side Feature Flags". After all, this is the main difference to the previous article, so it should be discussed at length.

It would be good to explain why LaunchDarkly provides a separate client-side key. The main reason is security, I guess, because a client must not know all the details of feature flags (including user IDs, feature flag rules, etc). See the docs here: https://docs.launchdarkly.com/sdk/concepts/client-side-server-side

Also, this text is almost 1:1 copied from the LaunchDarkly docs. Please use your own words to describe this.

Reply:
I have created a separate section, "Differences Between Client-side and Server-side Feature Flags". Please review if that looks good.